doc/sphinx/user-extensions/syntax-extensions.rst

```diff
@@ -7,64 +7,177 @@ In this chapter, we introduce advanced commands to modify the way Coq
 parses and prints objects, i.e. the translations between the concrete
 and internal representations of terms and commands.
 
-The main commands to provide custom symbolic notations for terms are
-:cmd:`Notation` and :cmd:`Infix`; they will be described in the
-:ref:`next section <Notations>`. There is also a
-variant of :cmd:`Notation` which does not modify the parser; this provides a
-form of :ref:`abbreviation <Abbreviations>`. It is
-sometimes expected that the same symbolic notation has different meanings in
-different contexts; to achieve this form of overloading, Coq offers a notion
-of :ref:`notation scopes <Scopes>`.
-The main command to provide custom notations for tactics is :cmd:`Tactic Notation`.
+When inputing a term to Coq, it goes through several successive steps.
+
+* lexing: the input is split in a linear stream of tokens, for
+  instance the string ``"1 + 2 * 2"`` will be split into the sequence
+  of five tokens `'1' '+' '2' '*' '2'`
+
+* parsing: the previous stream is given a tree-like structure, for
+  instance here
```

Comment:
The parser doesn't actually construct a parse tree data structure.  Rather, as each grammar production is recognized (corresponding to non-leaf nodes in the parse tree), the parser calls the corresponding action function.  The tree is conceptually useful for thinking about, but better to be clear that it's just a concept, not the implementation.

Reply:
That's not actually true. Keep in mind, that this chapter is not about the parser in general, camlp5, mlg files or what not. It's only about the notation mechanism. There is indeed some kind of action function inside the notation mechanism but its only role is to build the tree. So I don't think there is any point mentioning this internal detail to the reader here.

Reply:
The parser does not build a parse tree data structure.  (I'm quite familiar with the parsing code.)  Would you point me to the code that you believe "gives a tree-like structure"?  Are you thinking of some code specific to notations?

I suggest tweaking the wording so it's clear that the tree is a (useful) concept and not suggest there is a tree data structure.  When the doc gives a simplified explanation of something, it should still be precise and accurate as far as it goes.

> So I don't think there is any point mentioning this internal detail to the reader here.

Agreed, however descriptions generally need to take into account much more detail than will appear in the doc.  I certainly wouldn't mention action routines here, but their use influences what is said.

Reply:
> The parser does not build a parse tree data structure. (I'm quite familiar with the parsing code.)

Bear in mind that this chapter of the refman is not about the generic parser but about the specifics of the Notation mechanism. And while, indeed, generic parsers could, in theory, do anything else than building an AST, they usually do and in our particular case it definitely does (and does nothing else).

> Would you point me to the code that you believe "gives a tree-like structure"? Are you thinking of some code specific to notations?

https://github.com/coq/coq/blob/a7f51315db5d70888af3b96a579eb799a2b45ca9/interp/constrexpr.mli#L170

(and the rule building it is basically the `make_act` function in egramcoq.ml)

> I suggest tweaking the wording so it's clear that the tree is a (useful) concept and not suggest there is a tree data structure. When the doc gives a simplified explanation of something, it should still be precise and accurate as far as it goes.

No, as shown above, the tree definitely exists, it's not a simplified inacurate explanation by any mean. And I stand by the fact that it being built by some internal camlp5 rule (the `make_act` function mentionned above) is a technical detail of absolutely no interest for the reader.